src/Storages/MergeTree/DataPartStorageOnDisk.cpp

```diff
@@ -322,12 +324,19 @@ void DataPartStorageOnDisk::clearDirectory(
     const NameSet & names_not_to_remove,
     const MergeTreeDataPartChecksums & checksums,
     const std::unordered_set<String> & skip_directories,
+    bool is_temp,
+    MergeTreeDataPartState state,
     Poco::Logger * log,
     bool is_projection) const
 {
     auto disk = volume->getDisk();
 
-    if (checksums.empty())
+    /// It does not make sense to try fast path for incomplete temporary parts, because some files are probably absent.
+    /// Sometimes we add something to checksums.files before actually writing checksums and columns on disk.
```

Comment:
For me, it's unclear which exact cases can lead to such a situation.

Reply:
Just because code is written that way. We add something to `checksums.files` before writing `checksums.txt` (and `columns.txt`): [link](https://github.com/ClickHouse/ClickHouse/blob/0492ef63a40673b457e72892cfc3226245a41f65/src/Storages/MergeTree/MutateTask.cpp#L608-L649). So if exception is thrown between these points, then `checksums.empty()` is false, but some file on disk does not exist, so fast path will fail. And [here](https://github.com/ClickHouse/ClickHouse/blob/4cf03538921f002d4dcc137cbfda11bdba4d2bab/src/Storages/MergeTree/MergedBlockOutputStream.cpp#L272-L293) is an example of creating `checksums.txt` and `columns.txt` in different order (so we have to check both)